recreate.go

```diff
@@ -0,0 +1,442 @@
+// +build !cassandra
+// Copyright (C) 2017 ScyllaDB
+
+package gocql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"text/template"
+)
+
+// ToCQL returns a CQL query that ca be used to recreate keyspace with all
+// user defined types, tables, indexes, functions, aggregates and views associated
+// with this keyspace.
+func (km *KeyspaceMetadata) ToCQL() (string, error) {
+	sb := &strings.Builder{}
+
+	if err := km.keyspaceToCQL(sb); err != nil {
+		return "", err
+	}
+
+	sortedTypes := km.typesSortedTopologically()
+	for _, tm := range sortedTypes {
+		if err := km.userTypeToCQL(sb, tm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, tm := range km.Tables {
+		if err := km.tableToCQL(sb, km.Name, tm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, im := range km.Indexes {
+		if err := km.indexToCQL(sb, im); err != nil {
+			return "", err
+		}
+	}
+
+	for _, fm := range km.Functions {
+		if err := km.functionToCQL(sb, km.Name, fm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, am := range km.Aggregates {
+		if err := km.aggregateToCQL(sb, am); err != nil {
+			return "", err
+		}
+	}
+
+	for _, vm := range km.Views {
+		if err := km.viewToCQL(sb, vm); err != nil {
+			return "", err
+		}
+	}
+
+	return sb.String(), nil
+}
+
+func (km *KeyspaceMetadata) typesSortedTopologically() []*TypeMetadata {
+	sortedTypes := make([]*TypeMetadata, 0, len(km.Types))
+	for _, tm := range km.Types {
+		sortedTypes = append(sortedTypes, tm)
+	}
+	sort.Slice(sortedTypes, func(i, j int) bool {
+		for _, ft := range sortedTypes[j].FieldTypes {
+			if strings.Contains(ft, sortedTypes[i].Name) {
+				return true
+			}
+		}
+		return false
+	})
+	return sortedTypes
+}
+
+func (km *KeyspaceMetadata) tableToCQL(w io.Writer, kn string, tm *TableMetadata) error {
+	t := template.Must(template.New("table").
```

Comment:
Can the templates be global - they are safe for concurent use.

The functions they include do not have to be methods of KeyspaceMetadata.